FEA add newton-lsmr solver to LogisticRegression and GLMs

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Supersedes #23507.
Fixes #16634.

#### What does this implement/fix? Explain your changes.
This PR adds a further `NewtonSolver`: `NewtonLSMRSolver`.
This solver uses the iteratively reweighted least squares (IRLS) formulation of a Newton step. This means the inner solver uses the square root of the Hessian and solves the corresponding least squares problem (as opposed to solving the normal equation as `"newton-cholesky"` is doing) with the iterative [LSMR](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lsmr.html) solver.

This solver is therefore suited for dense and sparse `X`.

#### Any other comments?
The multinomial/multiclass case deserves special attention as there are different ways to look at the Hessian $X' W X$:
- Naive: Use, e.g. for `n_classes=3`
  ```
  X = [X 0 0]  W = [W00 W10 W00]  y = [y==0    0    0]  coef = [coef_class_0]  ...
      [0 X 0]      [W10 W11 W00]      [   0 y==1    0]         [coef_class_1]
      [0 0 X]      [W20 W10 W22]      [   0    0 y==2]         [coef_class_2]
  ```
- Consider every 1d-array/2d-array as a 2d-/3d-array with its 2nd/3rd dimension having all `n_samples`. Then
  $W = \mathrm{diag}(p) - p'p$ and $p$ the probability array in n_classes (and n_samples as "depth").
  Now, one can use the LDL decomposition of this particular matrix $W$, analytically given in https://doi.org/10.1111/j.2517-6161.1992.tb01875.x, use $\sqrt{D} L'$ as square root of $W$.
  This is the chosen approach.

Edit: For benchmarks, see https://github.com/scikit-learn/scikit-learn/pull/25462#issuecomment-1575538110.

lorentzenchr committed:
ENH add NewtonLSMRSolver

lorentzenchr committed:
TST add test_solver_on_ill_conditioned_X

lorentzenchr committed:
ENH add multinomial to LSMR

lorentzenchr committed:
ENH add newton-lsmr to LogisticRegression

lorentzenchr commented:
@ogrisel @mathurinm @TomDLT @agramfort @rth might be interested as this seems to be new ground for GLM solvers, especially the multinomial logistic regression!

It was a very stony path to arrive with all (added) tests green. Right now, I've no energy to do extensive benchmarking. But I hope, that this work will become useful and find its way into scikit-learn, in the end. I'm sure, there are opportunities left for performance optimization.

ogrisel commented:
Glad to see this! I just re-ran the previous benchmark for Poisson regression on the French MTPL dataset from the previous PR:

- https://github.com/scikit-learn/scikit-learn/pull/23507#issuecomment-1144003707

and here are the results on my laptop:

![poisson_reg_MTPL_iter](https://user-images.githubusercontent.com/89061/214252652-ae684fa8-f608-4c31-a4aa-132f75bb2f88.png)

![poisson_reg_MTPL](https://user-images.githubusercontent.com/89061/214252647-15fc1186-b68a-43b3-86b2-22d36b02e21f.png)

so this looks very good.

ogrisel commented on sklearn/linear_model/_glm/_newton_solver.py:
```diff
@@ -516,3 +532,460 @@ def inner_solve(self, X, y, sample_weight):
             return
+
+
+class NewtonLSMRSolver(NewtonSolver):
```
I have the feeling the code would be simpler to follow if we had a special subclass to handle the `if self.linear_loss.base_loss.is_multiclass` case.

That would require introducing a factory function to do the dispatching to the correct solver class based on the loss object but I have the feeling that would be worth it.

ogrisel commented on sklearn/linear_model/_glm/_newton_solver.py:
```diff
@@ -516,3 +532,460 @@ def inner_solve(self, X, y, sample_weight):
+        G = X.T @ g + l2_reg_strength * coef
+        H = X.T @ diag(h) @ X + l2_reg_strength * P
+        g = loss.gradient = pointwise gradient
+        h = loss.hessian = pointwise hessian
```
I think "Hessian" should always be capitalized in comments and docstrings.

```suggestion
        h = loss.hessian = pointwise Hessian
```

lorentzenchr replied:
In a loooot of other places, we simply use "hessian" not "Hessian".

ogrisel commented on sklearn/linear_model/_glm/_newton_solver.py:
```diff
@@ -516,3 +532,460 @@ def inner_solve(self, X, y, sample_weight):
+        # self.A_norm = np.sqrt(A_norm)
+        # As this all is a bit costly, we make life easier and start with 1:
+        self.A_norm = 1
+        self.r_norm = 1
```
Why is this valid?

lorentzenchr replied:
Because it works😏 Without penalty and with $h=1$, one has $||A|| = ||X||$. We likely have $||X|| > 1$, so `atol ~ 1/||A||` is a looser criterion with an underestimated $||A||$ which is a good choice for the first iteration.

lorentzenchr replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/25462/commits/59c13221b94c3e058d087446aa949deb3f049f83.

ogrisel reviewed (commented):
Some early feedback after a quick glance at the code.

Out of curiosity, have you tried to profile this to pinpoint the bottlenecks for both the multinomial and non-multinomial cases?

I have the impression that multithreading is barely used (in the multiclass case) so it's probably not a few large BLAS calls as is the case for `newton-cholesky`.

lorentzenchr committed:
CLN fix dtype and tests

- keep dtype float32 after LSMR
- lower test precision in
 test_NewtonLSMRSolver_multinomial_A_b_on_3_classes

lorentzenchr commented:
> Out of curiosity, have you tried to profile this to pinpoint the bottlenecks for both the multinomial and non-multinomial cases?

For the multinomial/multiclass case, it clearly is `LDL.sqrt_D_Lt_matmul` and `LDL.L_sqrt_D_matmul` in `A.matmul` and `A.rmatmul` inside LSMR.

lorentzenchr committed:
ENH speed up LDL by better handling q==0

lorentzenchr committed:
ENH speed up LDL by using einsum and q_inv

ogrisel commented:
@lorentzenchr I am not sure if you saw but your last push triggered a CI failure. I did not investigate myself but I wanted to make sure that it not go unnoticed.

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

lorentzenchr committed:
TST fix atol in test_multinomial_identifiability_properties

lorentzenchr commented:
Looking at the last plot, I wonder why the LMSR-based solver seems to slow down after the first 4 iterations, before it accelerates in the last two again. Perhaps, the choice of `atol` in `lsmr` in the `inner_solve` could be improved.

lorentzenchr committed:
DOC add whatsnew

lorentzenchr commented:
The remaining CI error will be automatically fixed with setting scipy>=1.4, see https://github.com/scipy/scipy/issues/7396.

ogrisel commented:
> The remaining CI error will be automatically fixed by setting scipy>=1.4, see https://github.com/scipy/scipy/issues/7396.

For reference, the bump to scipy>=1.4 in `main` happens in this PR: #24665

lorentzenchr committed:
TST reduce tolerances

lorentzenchr committed:
TST skip LinearOperator transpose for scipy<1.4

lorentzenchr committed:
TST fix skipif

lorentzenchr committed:
TST loosen rtol a bit

lorentzenchr commented:
CI all 🟢 again.

lorentzenchr committed:
TST make tests pass for all random seeds

* SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -x

lorentzenchr commented:
If we want, we could go on and optimize with multi-threaded Cython code like
```Cython
def sqrt_D_Lt_matmul_cython(
    double[::1, :] x,
    const double[::1, :] p,
    const double[::1, :] q_inv,
    int n_threads=1,
):
    cdef:
        int n_classes = x.shape[1]
        int n_samples = x.shape[0]
        int i, j, k
    for i in range(0, n_classes - 1):  # row i
        # L_ij = -p_i / q_j, we need transpose L'
        for j in range(i + 1, n_classes):  # column j
            for k in prange(n_samples, schedule='static', nogil=True, num_threads=n_threads):
                x[k, i] -= p[k, j] * q_inv[k, i] * x[k, j]
    return np.asarray(x)
``` 
This gives a little speed up, but I did not measure the impact on `solve`. The other bottleneck is `coef @ X.T` and making this multi-threaded is harder as it calls BLAS routines (dense X) or scipy.sparse routines (scipy decided against openmp).

I think, this solver is still a valuable addition.

lorentzenchr committed:
DOC add comment about initial A_norm

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

lorentzenchr commented:
Some more benchmark with many more classes using `make_classification`: `n_classes=n_features=50`, `n_samples=100_000` and dense `X`.

<details>

```python
import warnings
from pathlib import Path
import numpy as np
from scipy import sparse
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.exceptions import ConvergenceWarning
from sklearn.linear_model._linear_loss import LinearModelLoss
from sklearn.metrics import log_loss
from sklearn.model_selection import train_test_split
from time import perf_counter
import pandas as pd


X, y = make_classification(
    n_samples=200_000,
    n_features=50,
    n_informative=50,
    n_redundant=0,
    n_repeated=0,
    n_classes=50,
    random_state=42,
)


# X = X.toarray()
X_train, X_test, y_train, y_test = train_test_split(
    X, y, train_size=0.5, test_size=0.5, random_state=0
)
print(f"{X_train.shape = }")
print(f"{sparse.issparse(X_train)=}")
print(f"n_classes = {len(np.unique(y))}")

results = []
slow_solvers = set()
for tol in np.logspace(-1, -10, 10):
    for solver in ["lbfgs", "newton-cg", "newton-lsmr"]:
        if solver in slow_solvers:
            # skip slow solvers to keep the benchmark runtime reasonable
            continue
        tic = perf_counter()
        # with warnings.catch_warnings():
        #     warnings.filterwarnings("ignore", category=ConvergenceWarning)
        clf = LogisticRegression(
            C=1e12, solver=solver, tol=tol, max_iter=10000
        ).fit(X_train, y_train)
        toc = perf_counter()
        train_time = toc - tic
        if train_time > 200:
            # skip this solver from now on...
            slow_solvers.add(solver)
        # TODO: handle the regularization term...
        train_loss = log_loss(y_train, clf.predict_proba(X_train))
        n_iter = clf.n_iter_[0]
        result = {
            "solver": solver,
            "tol": tol,
            "train_loss": train_loss,
            "train_time": train_time,
            "train_score": clf.score(X_train, y_train),
            "test_score": clf.score(X_test, y_test),
            "n_iter": n_iter,
            "converged": n_iter < clf.max_iter,
        }
        print(result)
        results.append(result)


results = pd.DataFrame.from_records(results)
filepath = Path().resolve() / "bench_multinomial_logistic_regression_mtpl.csv"
results.to_csv(filepath)


import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt


results = pd.read_csv(filepath)
results["suboptimality"] = results["train_loss"] - results["train_loss"].min() + 1e-15
fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="n_iter", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by iterations")

fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="train_time", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by time")
plt.show()
```

</details>

lorentzenchr committed:
FIX warning condition and cleaner for loop

lorentzenchr committed:
ENH improve Multinomial_LDL_Decomposition by precomputation

lorentzenchr committed:
DOC improve docstrings and comments

lorentzenchr committed:
ENH inner stopping criterion for LSMR with forcing sequence

lorentzenchr commented:
After an important improvement in commit e7368e775c5fb5ea34f18022c0141b218d0212a7 that sets a better inner stopping criterion with a forcing sequence like in Newton CG, the benchmarks are looking much better, even impressive!

## Poisson Regression with freMTPL2freq dataset
`n_features=75`, `n_samples=54_2410` (80% for the training data)
![image](https://github.com/scikit-learn/scikit-learn/assets/15324633/bdb22529-6f2f-48da-a28e-07c495545c6a)

**Code for reproducibility:**
<details>

```python
import warnings
from pathlib import Path
import numpy as np
from scipy import sparse
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.linear_model import PoissonRegressor
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder
from sklearn.preprocessing import StandardScaler, KBinsDiscretizer
from sklearn.exceptions import ConvergenceWarning
from sklearn.linear_model._linear_loss import LinearModelLoss
from sklearn.metrics import mean_poisson_deviance
from sklearn.model_selection import train_test_split
from time import perf_counter
import pandas as pd
import joblib


m = joblib.Memory(location=".", verbose=0)


@m.cache
def prepare_data():
    df = fetch_openml(data_id=41214, as_frame=True, parser='auto').frame
    df["Frequency"] = df["ClaimNb"] / df["Exposure"]
    log_scale_transformer = make_pipeline(
        FunctionTransformer(np.log, validate=False), StandardScaler()
    )
    linear_model_preprocessor = ColumnTransformer(
        [
            ("passthrough_numeric", "passthrough", ["BonusMalus"]),
            (
                "binned_numeric",
                KBinsDiscretizer(n_bins=10, subsample=None),
                ["VehAge", "DrivAge"],
            ),
            ("log_scaled_numeric", log_scale_transformer, ["Density"]),
            (
                "onehot_categorical",
                OneHotEncoder(),
                ["VehBrand", "VehPower", "VehGas", "Region", "Area"],
            ),
        ],
        remainder="drop",
    )
    y = df["Frequency"]
    w = df["Exposure"]
    X = linear_model_preprocessor.fit_transform(df)
    return X, y, w


X, y, w = prepare_data()

X_train, X_test, y_train, y_test, w_train, w_test = train_test_split(
    X, y, w, train_size=0.8, test_size=0.2, random_state=0
)
print(f"{X_train.shape = }")
print(f"{sparse.issparse(X_train)=}")

results = []
slow_solvers = set()
for tol in np.logspace(-1, -10, 10):
    for solver in ["lbfgs", "newton-cholesky", "newton-lsmr"]:
        if solver in slow_solvers:
            # skip slow solvers to keep the benchmark runtime reasonable
            continue
        tic = perf_counter()
        # with warnings.catch_warnings():
        #     warnings.filterwarnings("ignore", category=ConvergenceWarning)
        clf = PoissonRegressor(
            alpha=1e-12, solver=solver, tol=tol, max_iter=10000
        ).fit(X_train, y_train, sample_weight=w_train)
        toc = perf_counter()
        train_time = toc - tic
        if train_time > 200:
            # skip this solver from now on...
            slow_solvers.add(solver)
        # TODO: handle the regularization term...
        train_loss = mean_poisson_deviance(y_train, clf.predict(X_train), sample_weight=w_train)
        n_iter = clf.n_iter_
        result = {
            "solver": solver,
            "tol": tol,
            "train_loss": train_loss,
            "train_time": train_time,
            "train_score": clf.score(X_train, y_train),
            "test_score": clf.score(X_test, y_test),
            "n_iter": n_iter,
            "converged": n_iter < clf.max_iter,
        }
        print(result)
        results.append(result)


results = pd.DataFrame.from_records(results)
filepath = Path().resolve() / "bench_poisson_regression_mtpl.csv"
results.to_csv(filepath)


import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt


results = pd.read_csv(filepath)
results["suboptimality"] = results["train_loss"] - results["train_loss"].min() + 1e-20
fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="n_iter", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by iterations")

fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="train_time", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by time")
plt.show()
```

</details>

## Multinomial Logistic Regression
### freMTPL2freq dataset
`n_classes=12`, `n_features=75`, `n_samples=10_000` (subsample, dataset is larger)

#### Sparse X 
![image](https://github.com/scikit-learn/scikit-learn/assets/15324633/2a2e9609-bcbb-451e-be3a-1642f7ba0c20)

#### Dense X
![image](https://github.com/scikit-learn/scikit-learn/assets/15324633/b6177574-5aa9-4be0-bbe4-3cb81b9ffcb9)

**Code for reproducibility:**
<details>

```python
import warnings
from pathlib import Path
import numpy as np
from scipy import sparse
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder
from sklearn.preprocessing import StandardScaler, KBinsDiscretizer
from sklearn.exceptions import ConvergenceWarning
from sklearn.linear_model._linear_loss import LinearModelLoss
from sklearn.metrics import log_loss
from sklearn.model_selection import train_test_split
from time import perf_counter
import pandas as pd
import joblib


m = joblib.Memory(location=".", verbose=0)


@m.cache
def prepare_data():
    df = fetch_openml(data_id=41214, as_frame=True, parser='auto').frame
    df["Frequency"] = df["ClaimNb"] / df["Exposure"]
    log_scale_transformer = make_pipeline(
        FunctionTransformer(np.log, validate=False), StandardScaler()
    )
    linear_model_preprocessor = ColumnTransformer(
        [
            ("passthrough_numeric", "passthrough", ["BonusMalus"]),
            (
                "binned_numeric",
                KBinsDiscretizer(n_bins=10, subsample=None),
                ["VehAge", "DrivAge"],
            ),
            ("log_scaled_numeric", log_scale_transformer, ["Density"]),
            (
                "onehot_categorical",
                OneHotEncoder(),
                ["VehBrand", "VehPower", "VehGas", "Region", "Area"],
            ),
        ],
        remainder="drop",
    )
    y = df["Frequency"]
    w = df["Exposure"]
    X = linear_model_preprocessor.fit_transform(df)
    return X, y, w


X, y_orig, w = prepare_data()

print("binning the target...")
binner = KBinsDiscretizer(
    n_bins=300, encode="ordinal", strategy="quantile", subsample=int(2e5), random_state=0
)
y = binner.fit_transform(y_orig.to_numpy().reshape(-1, 1)).ravel().astype(np.int32)

# X = X.toarray()
X_train, X_test, y_train, y_test, w_train, w_test = train_test_split(
    X, y, w, train_size=10_000, test_size=10_000, random_state=0
)
print(f"{X_train.shape = }")
print(f"{sparse.issparse(X_train)=}")
print("y_train.value_counts() :")
print(pd.Series(y_train).value_counts())

results = []
slow_solvers = set()
for tol in np.logspace(-1, -10, 10):
    for solver in ["lbfgs", "newton-cg", "newton-lsmr"]:
        if solver in slow_solvers:
            # skip slow solvers to keep the benchmark runtime reasonable
            continue
        tic = perf_counter()
        # with warnings.catch_warnings():
        #     warnings.filterwarnings("ignore", category=ConvergenceWarning)
        clf = LogisticRegression(
            C=1e12, solver=solver, tol=tol, max_iter=10000
        ).fit(X_train, y_train)
        toc = perf_counter()
        train_time = toc - tic
        if train_time > 200:
            # skip this solver from now on...
            slow_solvers.add(solver)
        # TODO: handle the regularization term...
        train_loss = log_loss(y_train, clf.predict_proba(X_train))
        n_iter = clf.n_iter_[0]
        result = {
            "solver": solver,
            "tol": tol,
            "train_loss": train_loss,
            "train_time": train_time,
            "train_score": clf.score(X_train, y_train),
            "test_score": clf.score(X_test, y_test),
            "n_iter": n_iter,
            "converged": n_iter < clf.max_iter,
        }
        print(result)
        results.append(result)


results = pd.DataFrame.from_records(results)
filepath = Path().resolve() / "bench_multinomial_logistic_regression_mtpl.csv"
results.to_csv(filepath)


import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt


results = pd.read_csv(filepath)
results["suboptimality"] = results["train_loss"] - results["train_loss"].min() + 1e-15
fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="n_iter", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by iterations")

fig, ax = plt.subplots(figsize=(8, 6))
for label, group in results.groupby("solver"):
    group.sort_values("tol").plot(
        x="train_time", y="suboptimality", loglog=True, marker="o", label=label, ax=ax
    )
ax.set_ylabel("suboptimality")
ax.set_title("Suboptimality by time")
plt.show()
```

</details>

#### 20 newsgroup dataset (vectorized)
`n_classes=20`, `n_features=130_107`, `n_samples=18_846` and sparse `X`.
**TODO**

**Code for reproducibility:**
<details>

```python

```

</details>

### Conclusion
The LSMR-Newton solver can be used for very fast but rough estimates or for high precision estimates. For multinomial logistic regression, it seems superior to the current solvers.

@rth @TomDLT @agramfort @mathurinm ping in case you are interested.

ogrisel commented:
This looks very impressive. However it seems that there are missing results and code snippet for the 20 newsgroups case in the previous comment

lorentzenchr commented:
> This looks very impressive. However it seems that there are missing results and code snippet for the 20 newsgroups case in the previous comment

On purpose, note the TODO. The 20 newsgroup run takes a lot of time. And I don't find a *not fine-tuned* LSMR stopping criteria, that works well here. Using a larger penalization would make the problem go away, but the above benchmark I did is close to no penalization in an `n_feature > n_samples` setting: not so nice for optimizers.

lorentzenchr committed:
DOC improve docs and comments of Newton solvers

ogrisel reviewed (commented):
I started to take notes below for a first review but before going further, I have a question:

For GLMs with non-canonical link functions (e.g. Gamma and Tweedie with p > 1, the log link is not the canonical link) the Hessian is not equal to the FIM and therefore not guaranteed to be Positive Definite.

Is LSMR somehow robust to this? Or can this cause the solver to diverge? Or do we just fallback to LBFGS in that case?

Note: this cannot happen for logistic regression since the logit link (inverse logistic sigmoid) is canonical and the Hessian is Positive Definite as long as the predictions are not overconfident (exactly 0 or 1)

Note that this problem should already happen with `"newton-cholesky"`` without me realizing but the LFBGS fallback probably might hide it without me realizing. I wonder if you have specifically investigated this for Gamma regression with logit links @lorentzenchr.

EDIT: btw it reminds me of introducing a linear classifier fit with the Brier loss and the softmax link function that you did not quite like (https://github.com/scikit-learn/scikit-learn/issues/23525#issuecomment-1145091915) :)

jjerphan commented:
Thank you for those efforts, @lorentzenchr. I am happy that this new solver allows speeding up the resolution by several orders of magnitude both on suboptimality and train time!

I would like to review your work but I hardly have time to for now, unfortunately. :confused:

ogrisel commented:
Indeed I also tried 20 newsgroups and I confirm it is very slow with the current state of this PR.

> Edit: To give some more details, for n_samples > n_features it seems beneficial to have a short first Newton iteration with only a few inner iterations in LSMR. For the 20 newsgroup setting (with a tiiiiny or no penalization), if the first Newton iteration is stopped early (only a few inner iterations in LSMR), it is a least squares solution and then needs many Newton iterations. If one allows for many inner LSMR iterations in the first Newton iteration, it is solved as solution to Ax=b with much lower loss/objective function and then needs much less Newton iterations in the following.

I think we need to record the number of inner and out iterations and wallclock times for various problems (different shapes, conditioning, loss, regularization strength, sparsity patterns) + tols or maxiter for inner solver and low tols and large maxiter for the outer loop so that we can get a finer insight on a robust strategy to set the lsmr stopping conditions.

lorentzenchr commented:
> so that we can get a finer insight on a robust strategy to set the lsmr stopping conditions.

Honestly, we don’t have a good infrastructure for that and I spent weeks if not months running benchmarks for estimators, in particular GLMs.

For this particular solver, the n_samples > n_features setting is good, almost according to textbooks. A way to improve would be to introduce another stopping criteria for LSMR either by vendoring or by asking scipy (stop when `||A'r|| <= new_tol`, i.e. independent of `||r||` and `||A||`, see comments in code).
For the underparametrized case, it is also good as long as there is some L2 penalty. The above newsgroup example has practically no penalty and that setting is rarely ever covered for solvers or GLMs (I’m not aware of any publication). I‘m considering to make the initial `atol` smaller for exactly this setting, dependent on the penalty. A commit my follow with an update of the the above benchmarks (in the same github comment).

lorentzenchr committed:
ENH set initial A_norm based on n_samples, n_features and l2_reg_strength

lorentzenchr commented on sklearn/linear_model/_glm/_newton_solver.py:
```diff
@@ -194,6 +193,9 @@ def fallback_lbfgs_solve(self, X, y, sample_weight):
         self.coef = opt_res.x
+        if self.linear_loss.base_loss.is_multiclass:
+            n_classes = self.linear_loss.base_loss.n_classes
+            self.coef = self.coef.reshape((n_classes, -1), order="F")
```
I went with a comment, but for the time being without the pragma.

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
TST set lsmr to xfail in test_solver_on_ill_conditioned_X like other solvers

lorentzenchr committed:
Doc docstring solver arg in LogisticRegression

lorentzenchr commented:
> For GLMs with non-canonical link functions (e.g. Gamma and Tweedie with p > 1, the log link is not the canonical link) the Hessian is not equal to the FIM and therefore not guaranteed to be Positive Definite.
> 
> Is LSMR somehow robust to this? Or can this cause the solver to diverge? Or do we just fallback to LBFGS in that case?
> 
> Note: this cannot happen for logistic regression since the logit link (inverse logistic sigmoid) is canonical and the Hessian is Positive Definite as long as the predictions are not overconfident (exactly 0 or 1)
> 
> Note that this problem should already happen with `"newton-cholesky"`` without me realizing but the LFBGS fallback probably might hide it without me realizing. I wonder if you have specifically investigated this for Gamma regression with logit links @lorentzenchr.

All the glm tests with fixture `glm_dataset` also run a `GammaRegressor`. LSMR also falls back to LBFGS in certain cases, but I did not investigate or was not yet able to trigger such a case.

lorentzenchr committed:
ENH set btol back to self.tol

lorentzenchr committed:
TST higher tol in test_glm_regression_hstacked_X

lorentzenchr committed:
ENH increase conlim to make more tests pass

lorentzenchr commented:
Commit https://github.com/scikit-learn/scikit-learn/pull/25462/commits/83ce34fc7a528f096718bfff03f07d94253d3b0b passes `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -n auto -rfE sklearn/linear_model/_glm/tests/test_glm.py` locally on my laptop.

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

ogrisel commented:
When trying to trigger the LBFGS fallback by trying to make the model extremely confident so has to make the Hessian numerically degenerate I triggered the following ZeroDivisionError at iteration 26 instead:

```python
>>> import numpy as np
... from sklearn.linear_model import LogisticRegression
... 
... x = np.array([-1e24] * 1 + [1e24] * 1)
... X = x.reshape(-1, 1)
... y = (x > 0).astype(np.int32)
... 
... lr = LogisticRegression(solver="newton-lsmr", penalty=None, verbose=100).fit(X, y)
... lr.n_iter_
[...]
Newton iter=26
    norm(gradient) = 4170877078229.1255
Traceback (most recent call last):
  Cell In[2], line 8
    lr = LogisticRegression(solver="newton-lsmr", penalty=None, verbose=100).fit(X, y)
  File ~/code/scikit-learn/sklearn/base.py:1148 in wrapper
    return fit_method(estimator, *args, **kwargs)
  File ~/code/scikit-learn/sklearn/linear_model/_logistic.py:1321 in fit
    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, prefer=prefer)(
  File ~/code/scikit-learn/sklearn/utils/parallel.py:65 in __call__
    return super().__call__(iterable_with_config)
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/parallel.py:1085 in __call__
    if self.dispatch_one_batch(iterator):
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/parallel.py:901 in dispatch_one_batch
    self._dispatch(tasks)
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/parallel.py:819 in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/_parallel_backends.py:208 in apply_async
    result = ImmediateResult(func)
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/_parallel_backends.py:597 in __init__
    self.results = batch()
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/parallel.py:288 in __call__
    return [func(*args, **kwargs)
  File ~/mambaforge/envs/dev/lib/python3.11/site-packages/joblib/parallel.py:288 in <listcomp>
    return [func(*args, **kwargs)
  File ~/code/scikit-learn/sklearn/utils/parallel.py:127 in __call__
    return self.function(*args, **kwargs)
  File ~/code/scikit-learn/sklearn/linear_model/_logistic.py:485 in _logistic_regression_path
    w0 = sol.solve(X=X, y=target, sample_weight=sample_weight)
  File ~/code/scikit-learn/sklearn/linear_model/_glm/_newton_solver.py:426 in solve
    self.inner_solve(X=X, y=y, sample_weight=sample_weight)
  File ~/code/scikit-learn/sklearn/linear_model/_glm/_newton_solver.py:974 in inner_solve
    atol=eta * norm_G / (self.A_norm * self.r_norm),
ZeroDivisionError: float division by zero
```

We might need a `np.max(self.A_norm * self.r_norm, np.finfo(dtype).eps)` of similar here.

ogrisel commented:
Also note that `"newton-cholesky"` and `"lbfgs"` converge with `n_iter_ = [0]`.

ogrisel commented:
Actually there is a problem with lbfgs on the above problem, it does not converge:

```
ABNORMAL_TERMINATION_IN_LNSRCH                              

 Line search cannot locate an adequate point after MAXLS
  function and gradient evaluations.
  Previous x, f and g restored.
 Possible causes: 1 error in function or gradient evaluation;
                  2 rounding error dominate computation.
/Users/ogrisel/code/scikit-learn/sklearn/linear_model/_logistic.py:459: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
```

and

```
>>> lr.coef_
array([[0.]])
```

while `"newton-cholesky"` does converge on this toy yet extreme problem.

lorentzenchr committed:
ENH verbose print total number of LSMR iterations

lorentzenchr committed:
DOC enhance Taylor series and comments

lorentzenchr committed:
DOC inner_solve sets self.lsmr_iter

lorentzenchr committed:
ENH more robust choice of atol (with some memory)

lorentzenchr commented:
I opened #26707 for investigating the inner solver stopping criterion and run a log of benchmarks. There is no clear winner. I have to leave it as is: Either it is good enough in it's current shape or someone else needs to dig deeper.

My conclusion is that we have quite some room of improvement of the current solvers, like #24752. Also the `"newton-cg"` could likely be improved by doing what liblinear does, see Galli & Lin "A Study on Truncated Newton Methods for Linear Classification" (https://www.csie.ntu.edu.tw/~cjlin/papers/tncg/tncg.pdf or https://doi.org/10.1109/TNNLS.2020.3045836).
Currently, I'm not 100% convinced of the newton-lsmr, but it is such a nice solver for multiclass problems!

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

ogrisel commented:
> Currently, I'm not 100% convinced of the newton-lsmr, but it is such a nice solver for multiclass problems!

There are cases where it's indeed quite impressive based on the last benchmarks that are now collapsed in the discussion.

https://github.com/scikit-learn/scikit-learn/pull/25462#issuecomment-1575538110

But I agree that fixing #24752 would be helpful to get a clearer picture.

Also based on benchopt, it seems that SAG & SAGA are better reference solvers for 20 newsgroups, see e.g.:

https://benchopt.github.io/results/preprint_results_preprint_results_logreg_l2.html

I have no intuitions on why this would be the case.

You'll have to switch the dataset in the menu on the left to see the results on 20 newsgroups.

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

lorentzenchr committed:
FIX adapt to new loss and gradient scaling

lorentzenchr committed:
FIX sw_sum for newton-lsmr

lorentzenchr committed:
CLN fix import order

lorentzenchr committed:
Merge branch 'main'

lorentzenchr committed:
FIX tests with sw_sum

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

jjerphan reviewed (commented):
Some comments I wrote a few months ago (they might not all be relevant).

lorentzenchr commented on sklearn/linear_model/_linear_loss.py:
```diff
@@ -669,3 +677,247 @@ def hessp(s):
         return grad, hessp
+
+
+class Multinomial_LDL_Decomposition:
```
LDL seems a bit more common than LDLT, also scipy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.ldl.html.

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr_only

lorentzenchr commented on sklearn/linear_model/_linear_loss.py:
```diff
@@ -669,3 +677,247 @@ def hessp(s):
+        self.p = proba
+        q = 1 - np.cumsum(self.p, axis=1)  # contiguity of p
+        self.proba_sum_to_1 = proba_sum_to_1
+        if self.p.dtype == np.float32:
+            eps = 2 * np.finfo(np.float32).resolution
+        else:
+            eps = 2 * np.finfo(np.float64).resolution
```
I'm not 100% sure that `proba` and hence `self.p` is always either float32 or float64.

lorentzenchr commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -395,7 +406,7 @@ def test_consistency_path():
 
     # test for fit_intercept=True
-    for solver in ("lbfgs", "newton-cg", "newton-cholesky", "liblinear", "sag", "saga"):
+    for solver in set(SOLVERS) - set(["liblinear"]):
```
mistake

lorentzenchr commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -760,6 +771,8 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
             clf.set_params(
                 tol=1e-18, max_iter=10000, random_state=global_random_seed + 1
             )
+        elif solver == "newton-lsmr":
+            clf.set_params(tol=1e-6)
```
What do you have in mind?

lorentzenchr committed:
CLN address review comments, move to 1.5
